compiler-rt/lib/radsan/radsan.cpp

```diff
@@ -7,26 +7,137 @@
 */
 
 #include <radsan/radsan.h>
+
 #include <radsan/radsan_context.h>
+#include <radsan/radsan_flags.h>
 #include <radsan/radsan_interceptors.h>
-#include <unistd.h>
+
+#include "sanitizer_common/sanitizer_atomic.h"
+#include "sanitizer_common/sanitizer_common.h"
+#include "sanitizer_common/sanitizer_flag_parser.h"
+#include "sanitizer_common/sanitizer_flags.h"
+
+using namespace __sanitizer;
+
+namespace radsan {
+static __sanitizer::atomic_uint8_t radsan_inited{};
+static __sanitizer::atomic_uint8_t radsan_init_is_running{};
+static __sanitizer::atomic_uint64_t radsan_report_count{};
+static Flags radsan_flags{};
+
+Flags *flags() { return &radsan_flags; }
+
+void EnsureInitialized() {
```

Comment:
Naive question: why do we need this function at all if we know that `radsan_init` is called in the `.preinit` stage?

Reply:
naive answer: So I don't know anything about `.preinit`, and never encountered it before this project.

When I was sleuthing a bit, it seems as if this is only a linux thing:

https://github.com/llvm/llvm-project/blob/04c473bea3e0f135432698fcaafab52e1fe1b5ec/compiler-rt/lib/sanitizer_common/sanitizer_internal_defs.h#L95-L113

So in non-linux systems, there needs to be something that will initialize this functionality

Reply:
Much to my surprise, on OSX you can completely remove the call to `initialiseInterceptors` and the entire (original) `radsan_init` and things work fine! That call (in the original version of the code, in the radsan branch right now) is never called on mac.


So to clean up my answer from this now that I'm older and wiser: The `initialiseInterceptors` call does not need to be called on mac. It DOES need to be called on linux, and is taken care of in the `.preinit` array.

However, now we are doing a lot more in this function, making sure our `flags()` are set up, etc. To facilitate this, we need to ensure radsan is initialized before we use anything that needs it. Thus we have this complicated scheme that ensures it, and doesn't blow up if there is a data race.